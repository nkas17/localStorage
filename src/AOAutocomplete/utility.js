import React from 'react';
import _ from 'lodash';

export const getItemValue = item => item.caption;

// Unique key used to make elements unique when looping.
let key = 0;

/**
 * Renders the caption so that substrings matching what the user typed are bold
 * http://regexr.com/3g8tg
 *
 * @param {string} caption - string to render
 * @returns {Node[]} - array of strings, potentially with bold tags around some
 */
export const renderCaption = (caption, inputValue) => {
	const escapedInputValue = inputValue.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
	const findInputRegex = new RegExp(`(^${escapedInputValue})|(\\W)(${escapedInputValue})`, 'ig');
	const splitCaption = caption.split(findInputRegex);
	return _.reduce(splitCaption, (elementsSoFar, subCaption) => {
		if (_.isNil(subCaption)) {
			return elementsSoFar;
		}
		if (subCaption.toLowerCase() === inputValue.toLowerCase()) {
			return elementsSoFar.concat(<b key={++key}>{subCaption}</b>);
		}
		return elementsSoFar.concat(subCaption);
	}, []);
};

/**
 * Creates the default renderItem function which bolds
 * the parts of the caption matching inputValue.
 *
 * We use a higher-order function because we do not have control of the "input" in this scope
 *
 * @param {string} inputValue - value on the input tag
 */
export const renderItemGenerator = inputValue => (item, isHighlighted) => (
	<div
		key={++key}
		className={`arrf-autocomplete__item ${isHighlighted ? 'arrf-autocomplete__item--highlighted' : ''}`}
	>
		{renderCaption(item.caption, inputValue)}
	</div>
);

/**
 * Transforms the string to filter out unwanted formatting.
 * Currently changes the string to be lower case and replaces punctuation with spaces
 * @param {string} string
 * @returns {string} - the transformed string
 */
const _transformString = (string) => {
	const lowerCase = string.toLowerCase();
	const withoutPunctuation = lowerCase.replace(/[[\].,/#!$%^&*;:{}=\-_`~()\\]/g, ' ');
	return withoutPunctuation;
};
// / ^ *
// Items will appear in the menu if the value entered matches the beginning of a word from the caption or value
export const shouldItemRender = (item, currentValue) => {
	const transformedItemValue = _transformString(item.value);
	const transformedItemCaption = _transformString(item.caption);
	const transfoemdCurrentValue = _transformString(currentValue);
	return ` ${transformedItemValue} ${transformedItemCaption}`.toLowerCase().includes(` ${transfoemdCurrentValue}`);
};

// sort items by match position by default
// see https://github.com/reactjs/react-autocomplete/blob/master/lib/utils.js
export const sortItems = (a, b, value) => {
	const aLower = a.caption.toLowerCase();
	const bLower = b.caption.toLowerCase();
	const valueLower = value.toLowerCase();
	const queryPosA = aLower.indexOf(valueLower);
	const queryPosB = bLower.indexOf(valueLower);
	if (queryPosA !== queryPosB) {
		return queryPosA - queryPosB;
	}
	return aLower < bLower ? -1 : 1;
};

export const renderMenu = items => (
	<div className="arrf-autocomplete">{items}</div>
);

/**
 * Returns the caption to the item whose value property matches value.
 * If none are found returns the original value.
 *
 * @param {string} value - value to search for in the items
 * @param {object[]} items - array of items to query
 * @returns {string} - the caption to display
 */
export const getCaptionFromValue = (value, items) => {
	const itemWithValue = items.find(item => item.value === value);
	if (itemWithValue) {
		return itemWithValue.caption;
	}
	return value;
};

/**
 * Returns the value to the item whose caption property matches caption.
 * If none are found returns the original caption.
 *
 * @param {string} caption - caption to search for in the items
 * @param {object[]} items - array of items to query
 * @returns {string} - the value to store
 */
export const getValueFromCaption = (caption, items) => {
	const itemWithCaption = items.find(item => item.caption === caption);
	if (itemWithCaption) {
		return itemWithCaption.value;
	}
	return caption;
};
