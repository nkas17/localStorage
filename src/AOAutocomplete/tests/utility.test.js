import _ from 'lodash';
import { shouldItemRender, sortItems } from '../utility';

/** returns a function that calls shouldItemRender with an item and a fixed value */
const shouldItemRenderWithFixedValue = value => item => shouldItemRender(item, value);

describe('shouldItemRender', () => {
	it('returns all items when value is empty', () => {
		const items = [
			{ caption: 'test A', value: 'test A' },
			{ caption: 'test B', value: 'test B' },
			{ caption: 'test C', value: 'test C' },
		];
		const allTrue = _.every(items, shouldItemRenderWithFixedValue(''));
		expect(allTrue).toBe(true);
	});

	it('returns all items when value in all items', () => {
		const items = [
			{ caption: 'test A', value: 'test A' },
			{ caption: 'test B', value: 'test B' },
			{ caption: 'test C', value: 'test C' },
		];
		const allTrue = _.every(items, shouldItemRenderWithFixedValue('test'));
		expect(allTrue).toBe(true);
	});

	it('returns some items', () => {
		const items = [
			{ caption: 'a test', value: 'a test' },
			{ caption: 'another test', value: 'another test' },
			{ caption: 'b test', value: 'b test' },
		];
		const someTrue = _.some(items, shouldItemRenderWithFixedValue('a'));
		const allTrue = _.every(items, shouldItemRenderWithFixedValue('a'));
		expect(someTrue).toBe(true);
		expect(allTrue).toBe(false);
	});

	it('returns no items', () => {
		const items = [
			{ caption: 'test A', value: 'test A' },
			{ caption: 'test B', value: 'test B' },
			{ caption: 'test C', value: 'test C' },
		];
		const someTrue = _.some(items, shouldItemRenderWithFixedValue('d'));
		expect(someTrue).toBe(false);
	});

	it('returns items where value matches but caption does not', () => {
		const items = [
			{ caption: '', value: 'test A' },
			{ caption: '', value: 'test B' },
			{ caption: '', value: 'test C' },
		];
		const allTrue = _.every(items, shouldItemRenderWithFixedValue('test'));
		expect(allTrue).toBe(true);
	});

	it('returns items where caption matches but value does not', () => {
		const items = [
			{ caption: 'test A', value: '' },
			{ caption: 'test B', value: '' },
			{ caption: 'test C', value: '' },
		];
		const allTrue = _.every(items, shouldItemRenderWithFixedValue('test'));
		expect(allTrue).toBe(true);
	});

	it('returns items where some word starts with the value, but not items matching in the middle of a word', () => {
		const items = [
			{ caption: 'north carolina', value: '' },
			{ caption: 'oregon', value: '' },
		];
		const itemsToRender = _.filter(items, shouldItemRenderWithFixedValue('or'));
		expect(itemsToRender).toMatchSnapshot();
	});

	it('is case insensitive', () => {
		const items = [
			{ caption: 'CAPTION', value: '' },
			{ caption: 'caption', value: '' },
		];
		const allTrueLower = _.every(items, shouldItemRenderWithFixedValue('caption'));
		const allTrueUpper = _.every(items, shouldItemRenderWithFixedValue('CAPTION'));
		expect(allTrueLower).toBe(true);
		expect(allTrueUpper).toBe(true);
	});

	it('ignores punctuation', () => {
		const items = [
			{ caption: '[.,/#!$%^&*;:{}=-_`~()\\]Hello', value: '' },
		];
		const allTrue = _.every(items, shouldItemRenderWithFixedValue('Hello'));
		expect(allTrue).toBe(true);
	});
});

describe('sortItems', () => {
	it('returns alphabetical order when value has no match', () => {
		const itemA = { caption: 'A', value: 'A' };
		const itemB = { caption: 'B', value: 'B' };
		const firstItemLess = sortItems(itemA, itemB, '');
		const secondItemLess = sortItems(itemB, itemA, '');
		expect(firstItemLess).toBeLessThan(0);
		expect(secondItemLess).toBeGreaterThan(0);
	});

	it('returns positive when second arg has the earlier find position', () => {
		const itemA = { caption: 'north carolina', value: '' };
		const itemB = { caption: 'oregon', value: '' };
		const result = sortItems(itemA, itemB, 'or');
		expect(result).toBeGreaterThan(0);
	});

	it('returns negative when first arg has the earlier find position', () => {
		const itemA = { caption: 'oregon', value: '' };
		const itemB = { caption: 'north carolina', value: '' };
		const result = sortItems(itemA, itemB, 'or');
		expect(result).toBeLessThan(0);
	});
});
