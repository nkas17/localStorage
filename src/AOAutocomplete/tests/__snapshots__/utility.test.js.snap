// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shouldItemRender returns items where some word starts with the value, but not items matching in the middle of a word 1`] = `
Array [
  Object {
    "caption": "oregon",
    "value": "",
  },
]
`;
